File: JDK11/opentraceability/src/main/java/opentraceability/mappers/masterdata/GS1VocabJsonMapper.java
package opentraceability.mappers.masterdata;

import opentraceability.interfaces.*;
import opentraceability.utility.*;
import opentraceability.*;
import opentraceability.mappers.*;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.*;

public class GS1VocabJsonMapper implements IMasterDataMapper
{
	public String map(IVocabularyElement vocab) throws Exception {
		if (vocab.context == null)
		{
			vocab.context = new JSONObject("{" + "\r\n" + 
"                                    \"cbvmda\": \"urn:epcglobal:cbvmda:mda\"," + "\r\n" + 
"                                    \"xsd\": \"http://www.w3.org/2001/XMLSchema#\"," + "\r\n" + 
"                                    \"gs1\": \"http://gs1.org/voc/\"," + "\r\n" + 
"                                    \"@vocab\": \"http://gs1.org/voc/\"," + "\r\n" + 
"                                    \"gdst\": \"https://traceability-dialogue.org/vocab\"" + "\r\n" + 
"                                }");
		}

		Map<String, String> namespaces = GetNamespaces(vocab.context);
		Object obj = OpenTraceabilityJsonLDMapper.ToJson(vocab, DictionaryExtensions.reverse(namespaces));

		JSONObject json = obj instanceof JSONObject ? (JSONObject)obj : null;
		json.put("@context", vocab.context);
		return json.toString();
	}

	public final <T extends IVocabularyElement> IVocabularyElement map(String value, Class<T> clazz) throws Exception {
		return map(clazz, value);
	}

	public final IVocabularyElement map(java.lang.reflect.Type type, String value) throws Exception {
		JSONObject json = new JSONObject(value);

		JSONObject jContext = JSONExtensions.queryForObject(json,"@context");
		if (jContext == null)
		{
			throw new Exception("Failed to find @context.");
		}
		Map<String, String> namespaces = GetNamespaces(jContext);
		IVocabularyElement obj = (IVocabularyElement)OpenTraceabilityJsonLDMapper.FromJson(json, type, namespaces);
		obj.context = jContext;
		return obj;
	}

	private Map<String, String> GetNamespaces(Object jContext) throws Exception {

		// build our namespaces
		Map<String, String> namespaces = new HashMap<String, String>();
		if (jContext != null)
		{
			if (jContext instanceof JSONObject)
			{
				namespaces = JsonContextHelper.scrapeNamespaces((JSONObject)jContext);
			}
			else if (jContext instanceof JSONArray)
			{
				for (Object o : (JSONArray)jContext)
				{
					if (o instanceof JSONObject)
					{
						JSONObject j = (JSONObject) o;
						var ns = JsonContextHelper.scrapeNamespaces(j);
						for (var kvp : ns.entrySet())
						{
							if (!namespaces.containsKey(kvp.getKey()))
							{
								namespaces.put(kvp.getKey(), kvp.getValue());
							}
						}
					}
				}
			}
		}
		return namespaces;
	}
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/mappers/masterdata/gs1vocabjsonmapper.kt
package opentraceability.mappers.masterdata

import opentraceability.interfaces.IMasterDataMapper
import opentraceability.interfaces.IVocabularyElement
import opentraceability.mappers.OpenTraceabilityJsonLDMapper
import org.json.JSONArray
import opentraceability.utility.JsonContextHelper
import java.lang.reflect.Type
import java.util.*
import org.json.JSONObject;
import kotlin.reflect.KClass

class GS1VocabJsonMapper : IMasterDataMapper {
    override fun map(vocab: IVocabularyElement): String {
        if (vocab.context == null) {
            vocab.context = JSONObject("""{
                                    "cbvmda": "urn:epcglobal:cbvmda:mda",
                                    "xsd": "http://www.w3.org/2001/XMLSchema#",
                                    "gs1": "http://gs1.org/voc/",
                                    "@vocab": "http://gs1.org/voc/",
                                    "gdst": "https://traceability-dialogue.org/vocab"
                                }""")
        }

        val namespaces = getNamespaces(vocab.context ?: throw Exception("vocab.Context is null."))
        val reversedNamespaces = namespaces.entries.associate { (key, value) -> value to key }.toMutableMap()


        val json = OpenTraceabilityJsonLDMapper.toJson(vocab, reversedNamespaces) as JSONObject? ?: throw Exception("Failed to map master data into GS1 web vocab.")
        json.put("@context", vocab.context)
        return json.toString()
    }

    override fun <T : IVocabularyElement> map(type: KClass<T>, value: String): IVocabularyElement {
        val json = JSONObject(value)
        val namespaces = getNamespaces(json["@context"] ?: throw Exception("@context is null on the JSON-LD when deserializing GS1 Web Vocab. $value"))
        val obj = OpenTraceabilityJsonLDMapper.fromJson(json, type, namespaces) as IVocabularyElement
        obj.context = json.get("@context") as JSONObject
        return obj
    }

    override fun <T : IVocabularyElement> map(value: String): IVocabularyElement {
        return map<T>(value)
    }


    fun getNamespaces(jContext: Any): MutableMap<String, String> {
        val namespaces = mutableMapOf<String, String>()
        when (jContext) {
            is JSONObject -> {
                namespaces.putAll(JsonContextHelper.scrapeNamespaces(jContext))
            }
            is JSONArray -> {
                for (j in jContext) {
                    val ns = JsonContextHelper.scrapeNamespaces(j as JSONObject)
                    for ((k, v) in ns) {
                        if (!namespaces.containsKey(k)) {
                            namespaces[k] = v
                        }
                    }
                }
            }
        }
        return namespaces
    }
}
